### PR TITLE
[None][test] verify model feature matrix support

### PR DESCRIPTION
**Pull request by cascade812**
## Dev Engineer Review

- Adds shared KV-cache reuse assertions for deterministic output, output length, reuse metrics, and token equality.
- Refactors GLM52 and Kimi3 tests to use the shared helper.
- Extends DeepSeek-V3-Lite and GPT-OSS Eagle3 reuse coverage with explicit expected reuse counts.
- Updates model and feature matrices and removes obsolete `TLLM C++ Sampler` entries.
- Documents hybrid recurrent-attention snapshot requirements and known Glm4MoeLite logits limitations.
- CI follow-up remains required because the latest two runs failed.

## QA Engineer Review

- Modified `accuracy_core.py`, `test_glm52.py`, `test_kimi3.py`, and `test_llm_api_pytorch.py`.
- Added coverage for KV-cache reuse, attention-DP scheduling, chunked prefill, and GPT-OSS VSWA reuse.
- No test-list files changed. Existing `test-db/` entries and `waives.txt` contain related test references, but registration of all changed cases requires verification.
- Coverage verdict: **needs follow-up** because CI reported failures and several model cases remain unsupported or failed.

## Per-File QA Perspective

- `docs/source/models/supported-models.md`: Verify matrix statuses, language-only versus multimodal distinctions, and documented KV-cache and logits limitations.
- `docs/source/features/feature-combination-matrix.md`: Verify matrix alignment after removing the `TLLM C++ Sampler` row and column.
- `tests/integration/defs/accuracy/accuracy_core.py`: Verify shared KV-cache assertions and `SchedulingParams` handling.
- `tests/integration/defs/accuracy/test_glm52.py`: Verify attention-DP KV-cache reuse coverage through the shared helper.
- `tests/integration/defs/accuracy/test_kimi3.py`: Verify reuse-mode behavior after local-helper removal.
- `tests/integration/defs/accuracy/test_llm_api_pytorch.py`: Verify DeepSeek-V3-Lite and GPT-OSS expected reuse counts across scheduling and KV-cache configurations.

## Description

1. **KV cache reuse:** Add a shared `assert_kv_cache_reuse_for_llm` integration-test helper, use it in selected model accuracy tests, and update the model-feature matrix from architecture probes. Document the recurrent-state snapshot requirement for hybrid recurrent-attention models.
2. **Logits post processor:** Refresh model-feature matrix coverage from an end-to-end architecture sweep that forces generated tokens to token ID 22.
3. **TLLM C++ Sampler:** Remove the obsolete column from the language-model and multimodal support matrices and remove the deprecated feature from both axes of the feature-combination matrix.
4. **Multimodal matrix validation:** Clarify that language-only and multimodal requests cover different paths, validate overlapping architectures with real image inputs, and update the multimodal results independently.
5. **Gemma 4 runtime features:** Add parameterized end-to-end coverage for overlap scheduling, CUDA graphs, and four-way attention data parallelism, register it in the four-GPU GB200 pre-merge test list, and update the feature matrices.

### KV cache reuse

The local architecture probes call the checked-in `assert_kv_cache_reuse_for_llm` helper unless otherwise noted.

| Model architecture | Test result |
| --- | --- |
| `DeepseekV3ForCausalLM` | **PASS** — selected `TestDeepSeekV3Lite::test_chunked_prefill` configuration (`1 passed`) |
| `DeepseekV32ForCausalLM` | **PASS** — warm run reused 7 blocks; outputs identical |
| `GlmMoeDsaForCausalLM` | **PASS** — `TestGLM52NVFP4::test_runtime` (`1 passed`) |
| `DeepseekV4ForCausalLM` | **PASS** — 257-token TEP probe reused 4 blocks; outputs identical |
| `Glm4MoeForCausalLM` | **PASS** — direct cache-reuse assertion |
| `Qwen3MoeForCausalLM` | **PASS** — direct cache-reuse assertion |
| `Qwen3NextForCausalLM` | **PASS** — hybrid-state snapshot probe; outputs identical |
| `Qwen4ExpForCausalLM` | **Proxy PASS** — the conditional-generation checkpoint's inner text core compiled the FMHA H256 kernels and reused 8 blocks with identical output; no standalone checkpoint declaring this architecture was available |
| `Llama4ForConditionalGeneration` | **PASS** — direct cache-reuse assertion |
| `GptOssForCausalLM` | **PASS** — FMHA H64 kernels compiled; natural 256-token prompt reused 7 blocks; eight-token outputs identical |
| `Glm4MoeLiteForCausalLM` | **PASS** — AutoDeploy cache-reuse assertion |
| `NemotronHForCausalLM` | **PASS** — hybrid-state snapshot probe |
| `Step3p7ForConditionalGeneration` | **FAIL / unsupported** — warm run reused 8 blocks, but output tokens differed; matrix records `No` |
| `KimiK3ForConditionalGeneration` | **PASS** — existing cache-reuse assertion was mechanically moved to the shared helper |

### Logits post processor

Each executed test used `LLM.generate()` with a logits processor that forced four generated tokens to token ID 22.

| Model architecture | Test configuration | Result |
| --- | --- | --- |
| `DeepseekV3ForCausalLM` | Real checkpoint, H100 | **PASS** |
| `DeepseekV32ForCausalLM` | Real checkpoint, 8x B200, TP8/EP8 | **PASS** |
| `GlmMoeDsaForCausalLM` | Real checkpoint, 8x B200, TP8/EP8 | **PASS** |
| `DeepseekV4ForCausalLM` | Real checkpoint, 4x B200, TP4/EP4 | **PASS** |
| `Glm4MoeForCausalLM` | Real checkpoint, 4x B200, TP4/EP4 | **PASS** |
| `Qwen3MoeForCausalLM` | Not present in the supplied sweep results | Existing matrix status `Yes` retained |
| `Qwen3NextForCausalLM` | Real checkpoint, B200 | **PASS** |
| `Qwen3_5MoeForCausalLM` | Real checkpoint, 8x B300, TP8/EP8 | **PASS** |
| `Qwen4ExpForCausalLM` | Conditional-generation checkpoint, B200, TP4/EP4 | **Proxy PASS** — no exact text-only checkpoint available|
| `Llama4ForConditionalGeneration` | Real checkpoint, B200 | **PASS** |
| `GptOssForCausalLM` | Real checkpoint, H100 | **PASS** |
| `KimiK3ForConditionalGeneration` | Real checkpoint, 8x B300, TP8/EP8 | **PASS** |
| `Glm4MoeLiteForCausalLM` | Real checkpoint, AutoDeploy | **FAIL** — `ADEngine` lacks `_apply_logits_processors`; matrix records `No` |
| `NemotronHForCausalLM` | Real checkpoint, H100 | **PASS** |
| `Gemma4ForConditionalGeneration` | Real `nvidia-Gemma-4-26B-A4B-NVFP4` checkpoint, 1x B200, native PyTorch, FP8 KV cache with block/partial reuse disabled | **PASS**|
| `Gemma4UnifiedForConditionalGeneration` | Real `gemma-4-12B-it` checkpoint, image input, 1x B200, native PyTorch | **PASS**|
| `Step3p7ForConditionalGeneration` | Real FP8 checkpoint, image input, 4x B200, TP4/EP4 | **PASS**|
| `MiniMaxM3SparseForConditionalGeneration` | Real NVFP4 checkpoint, image input, 4x B200, TP4/EP4, MSA sparse path | **PASS** |

### Multimodal matrix validation

These tests exercised each architecture with at least one real image, rather than inferring multimodal support from its language-only decoder path.

| Model architecture | Feature | Test configuration | Result |
| --- | --- | --- | --- |
| `Gemma4UnifiedForConditionalGeneration` | Logits post processor | Real `gemma-4-12B-it`, 1x B200, native PyTorch | **PASS** |
| `Step3p7ForConditionalGeneration` | Logits post processor | Real FP8 checkpoint, 4x B200, TP4/EP4 | **PASS**|
| `MiniMaxM3SparseForConditionalGeneration` | Logits post processor | Real NVFP4 checkpoint, 4x B200, TP4/EP4, MSA sparse path | **PASS** |
| `Llama4ForConditionalGeneration` | KV cache reuse | Real `Llama-4-Maverick-17B-128E-Instruct-FP8`, 4x B200, TP4/EP4 | **FAIL / unsupported** — the reuse-disabled image baseline passed, while enabling block reuse failed with 0 image tokens for 720 image embeddings; multimodal matrix remains `No` |

### Gemma 4 runtime features

Each case uses the real checkpoint tokenizer and architecture config with dummy weights, eight decoder layers, and FlashInfer-supported attention head dimensions. Four identical requests are pinned one per attention-DP rank, and the test verifies equal four-token greedy outputs across ranks while CUDA graphs and overlap scheduling are enabled.

| Model architecture | Test configuration | Overlap scheduler | CUDA graph | Attention data parallelism |
| --- | --- | --- | --- | --- |
| `Gemma4ForConditionalGeneration` | 26B MoE config, 4x B200, TP4/EP4/ADP4 | **PASS** | **PASS** | **PASS** |
| `Gemma4UnifiedForConditionalGeneration` | 12B unified config, 4x B200, TP4/ADP4 | **PASS** | **PASS** | **PASS** |

Pre-commit hooks passed for all changed files.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by BowenFu — APPROVED** [review]

**Review by arysef — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

File: tests/integration/defs/accuracy/test_glm52.py [hunk 1]
```diff
@@ -250,7 +251,11 @@ def test_runtime(self, monkeypatch: pytest.MonkeyPatch) -> None:
             assert llm.args.kv_cache_config.enable_block_reuse is True
             assert llm.args.sparse_attention_config.skip_indexer_for_short_seqs is False
             self._assert_attention_dp(llm, num_ranks)
-            self._assert_kv_cache_reuse(llm)
+            assert_kv_cache_reuse_for_llm(
```
Comment (2ez4bz) [inline tests/integration/defs/accuracy/test_glm52.py:255]: Out of curiosity, how stable is this in practice?

File: tests/unittest/_torch/modeling/test_gemma4_e2e_dummy.py [hunk 2]
```diff
@@ -202,6 +202,8 @@ def test_e2e_text_26b_dummy():
             load_format="dummy",
             attn_backend="FLASHINFER",
             dtype="bfloat16",
+            disable_overlap_scheduler=False,
+            cuda_graph_config=CudaGraphConfig(),
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_gemma4_e2e_dummy.py]: Just checking - I thought cuda graphs were enabled by default? Was there something that necessitated this explicit line?

File: tests/integration/defs/accuracy/test_glm52.py (hunk 1 above)
Comment (cascade812) [reply tests/integration/defs/accuracy/test_glm52.py:255]: Based on my experience, this is quite stable, I tested it for all model architectures that appear in the feature support matrix, all models pass the test except that Step3p7ForConditionalGeneration has accuracy issue with kv cache reuse enabled.

Comment (2ez4bz) [reply tests/integration/defs/accuracy/test_glm52.py:255]: Thanks for confirming.

File: tests/unittest/_torch/modeling/test_gemma4_e2e_dummy.py (hunk 2 above)
Comment (cascade812) [reply tests/unittest/_torch/modeling/test_gemma4_e2e_dummy.py]: You’re right—CUDA graphs are enabled by default, will remove this redundant line.

**Review by BowenFu — APPROVED** [review]

**Review by jieli-matrix — APPROVED** [review]

**Review by Wanli-Jiang — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]
